examples/statemachine/src/language-server/statemachine-validator.ts

```diff
@@ -27,5 +32,29 @@ export class StatemachineValidator {
             }
         }
     }
+    /**
+     * Checks if there are duplicate state and event names.
+     * @param statemachine the statemachine to check
+     * @param accept the acceptor to report errors
+     */
+    checkUniqueStatesAndEvents(statemachine: Statemachine, accept: ValidationAcceptor): void {
+        const variables: string[] = [];
```

Comment:
There are two issues with this approach of checking duplicated elements. One is semantic in nature, the other is more technical.

1. We usually want to show both occurrences of duplicated elements. Right now the error is only raised on second element, never the first.
2. Using an array and `find` leads to [quadratic runtime complexity](https://jarednielsen.com/big-o-quadratic-time-complexity/). While it generally works, it's not a good showcase for users who are looking how to implement such a validation.

We have an optimized of this algorithm version available [here](https://github.com/langium/langium/blob/712bd01d58dc3c683ccbb27f813e1c0e7fc941f6/packages/langium/src/grammar/validation/validator.ts#L165-L189) which runs in linear time and addresses the first issue as well.